### PR TITLE
Limit paste creation / commenting to configurable IP addresses

### LINKED ISSUE #883: Usage Control
It would be a interesting function to allow the secret creation just for several ip adresses.
And on the other side, allow the retrieving for everyone with a valid link. 

I didn't found a solution for that one. 
Thanks!

### PR DESCRIPTION
**Pull request by elrido**
This PR fixes #883 and replaces #622.

## Changes
* Adds an option `creators` in `traffic` section to add a list of IPs (v4 or v6) or subnets (CIDR notation)
* If not empty, these creator IPs are the only ones allowed to create pastes or comment them, but anyone can still view the pastes.
* The original PR had placed this logic into the `Controller`, but when moved to `TrafficLimiter` the logic of the exempted IPs can be reused. Downside is that I now return true or an exception, so that a (translated) error message can be passed up to the `Controller`.
* Renamed the `exemptedIps` setting to `exempted` and changed the wording of both configuration items comments, so they match.
* For reference, the overall traffic limiting logic (which only applies when creating a paste or comment) now is as follows:
  * If `creators` isn't empty and the current IP matches one of the items on the list, let it pass, otherwise block it.
  * If the limit is set to less then 1 (= disabled), let it pass.
  * If the current (sha256 hashed) IP is in the list of limited IPs AND the limit hasn't expired, block it.
  * otherwise, let it pass.

In other words, once `creators` are defined, it is the only thing checked and anything else is skipped. So it is exclusive of disabling the limit or the `exempted` IPs.

## ToDo
* [x] Get this PR reviewed, incorporate any fixes or suggestions.
* [ ] Update the configuration settings wiki page.

### CODE REVIEW

**Comment by elrido** [comment]
FYI: Scrutinizer warns that in newer iplib versions a few functions got deprecated, so I suggest, we upgrade the libraries first, then update this and fix the deprecations, before merging it. Mark it as WIP.

**Comment by elrido** [comment]
library update is in, will get those deprecated function calls addressed on the weekend.

File: tst/Persistence/TrafficLimiterTest.php [hunk 1]
```diff
@@ -38,27 +38,74 @@ public function testTrafficGetsLimited()
...
+        $this->assertTrue(TrafficLimiter::canPass(), 'IPv4 in creator range');
+        $this->assertTrue(TrafficLimiter::canPass(), 'request is to fast, but IPv4 in creator range');
+        $_SERVER['REMOTE_ADDR'] = '2001:1620:2057:dead:beef::cafe:babe';
+        $this->assertTrue(TrafficLimiter::canPass(), 'IPv6 in creator range');
+        $this->assertTrue(TrafficLimiter::canPass(), 'request is to fast, but IPv6 in creator range');
+        TrafficLimiter::setCreators('127.*,foobar');
+        try {
+            $this->assertFalse(TrafficLimiter::canPass(), 'expected an exception');
+        } catch (Exception $e) {
+            $this->assertEquals($e->getMessage(), 'Your IP is not authorized to create pastes.', 'request is to fast, not a creator');
```
Comment (rugk) [inline tst/Persistence/TrafficLimiterTest.php:105]:
```suggestion
            $this->assertEquals($e->getMessage(), 'Your IP is not authorized to create pastes.', 'not a creator');
```

The error here is that the creator IP is not whitelisted, not that the request is too fast, or is it?

And if so, we should potentially wait some seconds before that, so the error is really due to not being whitelisted, not due to being a too fast request.

**Review by rugk — CHANGES_REQUESTED** [review]

**Comment by elrido** [comment]
deprecated function calls got fixed, pretty ready now, IMO

**Review by rugk — DISMISSED** [review]

**Comment by elrido** [comment]
Ah, sorry @rugk, I wasn't fast enough. Could you approve it once more, please? I merged master into this branch (with the corsican lang) and added the new message to the co.json and regenerated the SRI hashes for privatebin.js in the templates.

**Review by rugk — APPROVED** [review]

File: tst/Persistence/TrafficLimiterTest.php [hunk 2]
```diff
@@ -38,27 +38,74 @@ public function testTrafficGetsLimited()
...
+        TrafficLimiter::setCreators('1.2.3.4,10.10.10.0/24,2001:1620:2057::/48');
+        $_SERVER['REMOTE_ADDR'] = '127.0.0.1';
+        try {
+            $this->assertFalse(TrafficLimiter::canPass(), 'expected an exception');
+        } catch (Exception $e) {
+            $this->assertEquals($e->getMessage(), 'Your IP is not authorized to create pastes.', 'not a creator');
+        }
+        $_SERVER['REMOTE_ADDR'] = '10.10.10.10';
+        $this->assertTrue(TrafficLimiter::canPass(), 'IPv4 in creator range');
+        $this->assertTrue(TrafficLimiter::canPass(), 'request is to fast, but IPv4 in creator range');
```
[the remark this answers is not in the document]
Comment (elrido) [reply tst/Persistence/TrafficLimiterTest.php]: The traffic limiter only applies to paste creation and commenting, never to viewing. So, yes.
